File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
// TC- O(log(mn))
// SC- O(1)

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rows = matrix.size();
        int cols = matrix[0].size();

        // 2d matrix can be treated as a 1d sorted array here and all we need to do is binary search. given idx in array, the corresponding row = idx/cols, col = idx%cols;
        int left = 0;
        int right = rows*cols-1;
        while(left<=right){
            int mid = left+ ((right-left)/2);
            int r = mid/cols;
            int c= mid%cols;
            if(matrix[r][c]==target)
                return true;
            else if(matrix[r][c]<target)
                left++;
            else
                right--;
        }
        return false;
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
# TC- O(logmn)
# SC- O(1)

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:

        # matrix if read in row major order is already sorted, so binary search

        m, n = len(matrix), len(matrix[0])
        left, right = 0, m*n-1

        while(left<=right):
            
            mid = left + (right-left)//2
            r = mid // n
            c = mid % n

            if matrix[r][c]==target:
                return True
            
            elif target<matrix[r][c]:
                right = mid-1

            else:
                left = mid+1
        
        return False
            
